torch/optim/lr_scheduler.py

```diff
@@ -26,6 +43,7 @@ def __init__(self, optimizer, last_epoch=-1):
         self.base_lrs = list(map(lambda group: group['initial_lr'], optimizer.param_groups))
         self.last_epoch = last_epoch
```

Comment:
Previously, it was not possible for `self.last_epoch` to be set to `-1`. With your changes, it is now set to `-1` when you pass in the default. This change to the invariants of members of LRScheduler was not documented in the PR. Can you explain more to me why this is sound? Because the `last_epoch` field is not underscored, it counts as a public member field, and so this is a very clear BC-breaking change.

Reply:
The default value of `last_epoch` has not changed: it was and is `-1`. However, during the initialization, 

* now: `last_epoch` defaults to -1, `step` is called and increments `last_epoch` to 0. 
* before: `last_epoch` defaults to -1, changed to 0, and `step` is called and increments to 1.

All the tests though were calling `step` outside of initialization with the epoch parameter, thus overriding `last_epoch`. This indeed raises the question: is this change backward compatible? 

However, running test/test_optim.py with pytorch-nightly works, but leads to 5 test failures when removing `epoch` from `step` when you would expect the same answers as before.

```
diff --git a/test/test_optim.py b/test/test_optim.py
index e334962c9..2a365649a 100644
--- a/test/test_optim.py
+++ b/test/test_optim.py
@@ -565,7 +565,7 @@ class TestLRScheduler(TestCase):
 
         def old_pattern2():
             for e in range(epochs):
-                scheduler.step(e)
+                scheduler.step()
                 self.opt.step()
 
         self.assertWarnsRegex(old_pattern2, r'how-to-adjust-learning-rate')
@@ -599,7 +599,7 @@ class TestLRScheduler(TestCase):
 
         def old_pattern2():
             for e in range(epochs):
-                scheduler.step(e)
+                scheduler.step()
                 self.opt.step()
 
         self.assertWarnsRegex(old_pattern2, r'how-to-adjust-learning-rate')
@@ -627,7 +627,7 @@ class TestLRScheduler(TestCase):
 
         def old_pattern2():
             for e in range(epochs):
-                scheduler.step(e)
+                scheduler.step()
                 self.opt.step()
 
         self.assertWarnsRegex(old_pattern2, r'how-to-adjust-learning-rate')
@@ -1153,7 +1153,7 @@ class TestLRScheduler(TestCase):
         if isinstance(schedulers, _LRScheduler):
             schedulers = [schedulers]
         for epoch in range(epochs):
-            [scheduler.step(epoch) for scheduler in schedulers]
+            [scheduler.step() for scheduler in schedulers]
             for param_group, target in zip(self.opt.param_groups, targets):
                 self.assertAlmostEqual(target[epoch], param_group['lr'],
                                        msg='LR is wrong in epoch {}: expected {}, got {}'.format(
@@ -1184,7 +1184,7 @@ class TestLRScheduler(TestCase):
             targets.append([group['lr'] for group in self.opt.param_groups])
         self.setUp()
         for epoch in range(epochs):
-            scheduler.step(epoch)
+            scheduler.step()
             for i, param_group in enumerate(self.opt.param_groups):
                 self.assertAlmostEqual(targets[epoch][i], param_group['lr'],
                                        msg='LR is wrong in epoch {}: expected {}, got {}'.format(
@@ -1198,7 +1198,7 @@ class TestLRScheduler(TestCase):
                 if isinstance(scheduler, ReduceLROnPlateau):
                     scheduler.step(metrics[epoch])
                 else:
-                    scheduler.step(epoch)
+                    scheduler.step()
             if verbose:
                 print('epoch{}:\tlr={}'.format(epoch, self.opt.param_groups[0]['lr']))
             for param_group, target in zip(self.opt.param_groups, targets):
```

I would in fact argue that this fixes the initial stepping behavior. Thoughts?